src/web/src/app/space/service.ts

```diff
@@ -53,4 +148,245 @@ export class SpaceService {
+			expiredAt: expiredAt
+		});
+		return of(true)
+			.pipe(
+				delay(2000),
+				src => this.requestWatcher.WatchFor(src)
+			)
+	}
+
+	delInvitationById(invitationId: number): Observable<boolean> {
+		return of(true).pipe(
+			delay(2000),
+			switchMap(val => {
+				const index = this.invitations.findIndex(x => x.id === invitationId);
+				if (index < 0)
+				{
+					const err = new Error("Invalid Id");
+					return throwError(err);
+				}
+				this.invitations.splice(index, 1);
+				return of(val);
+			}),
+			src => this.requestWatcher.WatchFor(src)
+		);
+	}
+
+	delLinkById(linkId: string): Observable<boolean> {
+		return of(true).pipe(
+			delay(2000),
+			switchMap(val => {
+				const index = this.links.findIndex(x => x.id === linkId);
+				if (index < 0)
+				{
+					const err = new Error("Invalid Id");
+					return throwError(err);
+				}
+				this.links.splice(index, 1);
+				return of(val);
+			}),
+			src => this.requestWatcher.WatchFor(src)
+		);	
+	}
+
+	delSpaceById(spaceId: string): Observable<boolean> {
+		return of(true).pipe(
+			delay(2000),
+			switchMap(val => {
+				const index = this.spaces.findIndex(x => x.id === spaceId);
+				if (index < 0)
+				{
+					const err = new Error("Invalid Id");
+					return throwError(err);
+				}
+				this.spaces.splice(index, 1);
+				return of(val);
+			}),
+			src => this.requestWatcher.WatchFor(src)
+		);	
+	}
+
+	delUserById(userId: string): Observable<boolean> {
+		return of(true).pipe(
+			delay(2000),
+			switchMap(val => {
+				const index = this.users.findIndex(x => x.userId === userId);
+				if (index < 0)
+				{
+					const err = new Error("Invalid Id");
+					return throwError(err);
+				}
+				this.users.splice(index, 1);
+				return of(val);
+			}),
+			src => this.requestWatcher.WatchFor(src)
+		);	
+	}
+
+	fillInvitationDetailFields(invitations: SpaceInvitationDetail[]) {
+		let invitationsArray: SpaceInvitationDetail[] = [];
+		invitations.forEach(invite => invitationsArray.push(invite));
+
+		for (const invitation of invitationsArray) {
+			this.getById(invitation.spaceId).subscribe(space => {
+				for (const invitation of invitations) {
+					if (invitation.spaceId !== space.id)
+						continue;
+					invitation.spaceName = space.name;
+				}
+			});
+		}
+
+		const userService = this.injector.get(UserService);
+		
+		for (const invitation of invitationsArray) {
+			userService.getById(invitation.userId).subscribe(user => {
+				for (const invitation of invitations) {
+					if (invitation.userId !== user.id)
+						continue;
+					invitation.userLogin = user.login;
+				}
+			});
+		}
+	}
+
+	fillLinkDetailFields(links: SpaceLinkDetail[]) {
+		let linksArray: SpaceLinkDetail[] = [];
+		links.forEach(invite => linksArray.push(invite));
+
+		for (const link of linksArray) {
+			this.getById(link.spaceId).subscribe(space => {
+				for (const link of links) {
+					if (link.spaceId !== space.id)
+						continue;
+					link.spaceName = space.name;
+				}
+			});
+		}
+	}
```

Comment:
Тут **"<T extends SpaceFields & {spaceId: string}>"** означает что эта функция может принимать любой массив который содержит объекты в которых есть все поля из SpaceFields и ещё поле spaceId


```suggestion
	fillFields<T extends SpaceFields & {spaceId: string}>(items: T[]) {
		const ids = new Set<string>();
		items.forEach(item => ids.add(item.spaceId));
		for (const id of ids) {
			this.getById(id).subscribe(space => {
				for (const item of items) {
					if (item.spaceId !== space.id)
						continue;
					item.spaceName = space.name;
				}
			});
		}
	}
```

Reply:
А точно нужна такая одна универсальная функция? Ведь мы через неё будем дополнять spaceName для links и spaceName+userLogin для invitations. Как я понимаю, если мы захотим именно универсальную функцию, то нужно будет как-то в коде делать проверку на тип T, с которым функция работает в данный момент. И если это тип SpaceLinksDetail, то заполнять только spaceName, а если тип SpaceInvitationDetail, то дополнительно заполнять userLogin. Правильно ли я понимаю целевую логику?

Reply:
См. коммент выше. Там два вызова. Поля должен заполнять тот кто за них ответственен. SpaceService заполняет поля ServiceFields а UserService поля UserFields

Reply:
Понял, спасибо, дошло наконец-то))